src/lib.rs

```diff
@@ -6,52 +7,194 @@ extern crate rand;
+        };
+    }
+
+    fn should_continue(&self) -> bool {
+        return (self.valid_examples < self.max_examples)
+            && (self.valid_examples + self.invalid_examples < self.max_examples * 10)
+            && (self.interesting_examples == 0);
+    }
+
+    fn mark_finished(&mut self, source: DataSource, status: Status) -> () {
+        match status {
+            Status::Overflow => (),
+            Status::Valid => self.valid_examples += 1,
+            Status::Invalid => self.invalid_examples += 1,
+            Status::Interesting => {
+                self.interesting_examples += 1;
+                self.best_example = Some(Rc::new(source.record));
+            }
+        }
+    }
+
+    fn new_source(&mut self) -> DataSource {
+        return DataSource::from_random(self.random.gen());
+    }
+
+    fn failing_example(&self) -> Option<DataSource> {
+        match self.best_example {
+            None => return None,
+            Some(ref v) => return Some(DataSource::from_vec(v.clone())),
+        }
+    }
 }
 
 ruby! {
   class HypothesisCoreEngine {
     struct {
-      random: ChaChaRng,
       next_id: u64,
       children: HashMap<u64, DataSource>,
+      engine: Engine,
     }
 
-    def initialize(helix, seed: u64){
+    def initialize(helix, seed: u64, max_examples: u64){
       let xs: [u32; 2] = [seed as u32, (seed >> 32) as u32];
       HypothesisCoreEngine{
         helix,
-        random: ChaChaRng::from_seed(&xs),
         next_id: 0,
         children: HashMap::new(),
+        engine: Engine::new(max_examples, &xs),
       }
     }
 
+    def was_unsatisfiable(&self) -> bool {
+      return self.engine.valid_examples == 0 && self.engine.interesting_examples == 0;
+    }
+
     def new_source(&mut self) -> u64 {
       let result = self.next_id;
-      self.children.insert(result, DataSource{random: self.random.clone()});
-      self.random.next_u64();
+      self.children.insert(result, self.engine.new_source());
       self.next_id += 1;
       return result;
     }
 
+    def failing_example(&mut self) -> Option<u64> {
+      if let Some(source) = self.engine.failing_example() {
+        let result = self.next_id;
+        self.children.insert(result, source);
+        self.next_id += 1;
+        return Some(result);
+      } else {
+        return None;
+      }
+    }
+
+    def should_continue(&self) -> bool {
+      return self.engine.should_continue();
+    }
+
+    def finish_overflow(&mut self, id: u64){
+      mark_status_id(&mut self.engine, &mut self.children, id, Status::Overflow);
+    }
+
+    def finish_invalid(&mut self, id: u64){
+      mark_status_id(&mut self.engine, &mut self.children, id, Status::Invalid);
+    }
+
+    def finish_interesting(&mut self, id: u64){
+      mark_status_id(&mut self.engine, &mut self.children, id, Status::Interesting);
+    }
+
+    def finish_valid(&mut self, id: u64){
+      mark_status_id(&mut self.engine, &mut self.children, id, Status::Valid);
```

Comment:
The separation I'm trying to draw here is that Engine works directly with DataSource values and the weird id thing is only an artifact of the ruby class (it basically only exists because Helix won't let me sensibly wrap pointers to Rust types in Ruby objects).

(And this can't be made a function on the Rust struct for the Ruby class because AFAICT Helix doesn't support that).